Improve RDB to AOF conversion docs

threema-danilo opened the pull request:
As discussed in https://github.com/redis/redis/issues/12484, here is a suggestion for improved AOF conversion documentation.

The first commit just adds a simple warning. In the second commit, I removed the section for Redis <2.2, which is out of support for a long time now. (If I'm not mistaken, 2.2 was released in 2011.)

I can also squash the two commits at request.

sundb commented on docs/management/persistence.md:
```diff
@@ -259,42 +259,35 @@ and starts appending new data into the new file.
 when you restart the server the configuration changes will be lost and the
-server will start again with the old configuration.
-
-**Redis 2.0**
-
-* Make a backup of your latest dump.rdb file.
-* Transfer this backup into a safe place.
-* Stop all the writes against the database!
-* Issue a `redis-cli BGREWRITEAOF`. This will create the append only file.
-* Stop the server when Redis finished generating the AOF dump.
-* Edit redis.conf end enable append only file persistence.
-* Restart the server.
-* Make sure that your database contains the same number of keys it contained before the switch.
-* Make sure that writes are appended to the append only file correctly.
```
Delete by mistake?

sundb replied:
Not really, I still see some users still using redis 3.x.

threema-danilo replied:
Redis 3 was released 4 years after 2.0, and the current instructions will work from 2.2 onwards. But sure, I can re-add these instructions.

sundb commented on docs/management/persistence.md:
```diff
@@ -259,42 +259,35 @@ and starts appending new data into the new file.
 The first CONFIG command enables the Append Only File persistence.
 
-The second CONFIG command is used to turn off snapshotting persistence. This is optional, if you wish you can take both the persistence methods enabled.
+The second CONFIG command is used to turn off snapshotting persistence. This is
+optional, if you wish you can take both the persistence methods enabled.
```
We should avoid this change that just adding an extra line.

threema-danilo replied:
I did that for consistency with the rest of the file, but I can revert the change.

threema-danilo commented:
@sundb I made the requested changes and squashed the commits!

threema-danilo commented on docs/management/persistence.md:
```diff
@@ -259,26 +259,28 @@ and starts appending new data into the new file.
 
-There is a different procedure to do this in version 2.0 and later versions, as you
-can guess it's simpler since Redis 2.2 and does not require a restart at all.
+If you want to enable AOF in a server that is currently using RDB snapshots, you need to convert the data interactively using `redis-cli`.
```
I assume you mean this comment? https://github.com/redis/redis/pull/12506#discussion_r1300149787 I amended the commit!

oranagra commented on docs/management/persistence.md:
```diff
@@ -259,26 +259,28 @@ and starts appending new data into the new file.
+* Optionally disable RDB: `redis-cli config set save ""`
+* Make sure your database contains the same number of keys it contained previously.
 * Make sure writes are appended to the append only file correctly.
+* Update your `redis.conf` and ensure that it matches the config above.
```
maybe add a reference to CONFIG REWRITE?

threema-danilo replied:
Maybe like this?

> * Update your `redis.conf` (potentially through `CONFIG REWRITE`) and ensure that it matches the config above.

oranagra commented on docs/management/persistence.md:
```diff
@@ -259,26 +259,28 @@ and starts appending new data into the new file.
-* Make sure your database contains the same number of keys it contained.
+* Enable AOF: `redis-cli config set appendonly yes`
+* Optionally disable RDB: `redis-cli config set save ""`
+* Make sure your database contains the same number of keys it contained previously.
```
IMHO that line is excessive (and silly)

threema-danilo replied:
I simply kept that line, which was already part of the docs before.

In general, a quick check that your data is still fine isn't totally wrong. Maybe we could word it like this?

    Make sure your database still contains all data, e.g. `redis-cli keys '*'`

oranagra replied:
but it's silly since enabling AOF can't (technically) affect your data in redis.
maybe instead we can add a line that suggests to wait till that initial AOFRW process completes?
i.e. if you'll try to restart redis before that happens, it'll refuse to go down.

oranagra replied:
sadly, i don't see that we exposed the status of `server.aof_state` in INFO, so we may need to suggest to look at some combination of `aof_rewrite_in_progress`, `aof_rewrite_scheduled`, and maybe `aof_last_bgrewrite_status`.
WDYT?

threema-danilo replied:
> but it's silly since enabling AOF can't (technically) affect your data in redis.

That's what I thought, and then all data was gone after restarting Redis (and it took 4 hours for someone to notice) 😉 (I know what you mean though, from a technical point of view.)

> sadly, i don't see that we exposed the status of `server.aof_state` in INFO, so we may need to suggest to look at some combination of `aof_rewrite_in_progress`, `aof_rewrite_scheduled`, and maybe `aof_last_bgrewrite_status`. WDYT?

These are all keys that are listed when querying `INFO persistence`, correct? What would be the correct combination of keys? (I don't know enough about Redis internals to propose some wording.)

oranagra replied:
> > but it's silly since enabling AOF can't (technically) affect your data in redis.
> 
> That's what I thought, and then all data was gone after restarting Redis (and it took 4 hours for someone to notice) 😉 (I know what you mean though, from a technical point of view.)

but that line tells you to check the data before restart, when all you did is change a configuration that doesn't affect the data in any way (redis doesn't touch it), so it's silly.
i.e. it does affect the guarantees, and how that data is persisted, but that's not what that line suggests to check.
IMHO we should delete it.

> > sadly, i don't see that we exposed the status of `server.aof_state` in INFO, so we may need to suggest to look at some combination of `aof_rewrite_in_progress`, `aof_rewrite_scheduled`, and maybe `aof_last_bgrewrite_status`. WDYT?
> 
> These are all keys that are listed when querying `INFO persistence`, correct? What would be the correct combination of keys? (I don't know enough about Redis internals to propose some wording.)

how about:
```
You may want to wait for AOF rewrite to finish persisting the data. You can do that by watching `INFO persistence`, waiting for `aof_rewrite_in_progress` and `aof_rewrite_scheduled` to be `0`, and validate that `aof_last_bgrewrite_status` is `ok`.
```

zuiderkwast replied:
This (check that the database contains the same number of keys it contained previously) makes sense to do after restarting. That's where things can go wrong.

I think it is better to move this bullet to a step after all other steps "Next time you restart the server, check that..."

I think we shall also add Oran's suggestion. It is important especially if you want to restart redis just after changing the config, when AOF rewrite is ongoing. I think we should also add "Before restarting, " to Oran's suggestion, so:

> Before restarting, you may want to wait for AOF rewrite to finish persisting the data. You can do that by watching `INFO persistence`, waiting for `aof_rewrite_in_progress` and `aof_rewrite_scheduled` to be `0`, and validate that `aof_last_bgrewrite_status` is `ok`.

oranagra reviewed (approved):
two minor suggestions.
i'd like @itamarhaber to review if he can.

zuiderkwast commented on docs/management/persistence.md:
```diff
@@ -259,26 +259,28 @@ and starts appending new data into the new file.
+* Optionally disable RDB: `redis-cli config set save ""`
+* Make sure your database contains the same number of keys it contained previously.
 * Make sure writes are appended to the append only file correctly.
+* Update your `redis.conf` (potentially through `CONFIG REWRITE`) and ensure that it matches the config above.
```
A few lines below this, there is:

> **IMPORTANT:** remember to edit your redis.conf to turn on the AOF, otherwise
when you restart the server the configuration changes will be lost and the

We should combine these two since they're saying the same thing. Can we move the IMPORTANT into this bullet?

threema-danilo replied:
Yes, good point. I think the new structure (Preparations / Switch / After Restart) helps with that as well.

zuiderkwast reviewed (commented):
Good. It's important to get this merged so people don't do this again. It's good we now have a warning in redis.conf too. I have a few small comments though.

threema-danilo commented:
Rebased and updated! @zuiderkwast @oranagra I integrated your suggestions and grouped the steps into three groups (Preparations, Switch, After Restart). I think that helps.

zuiderkwast commented on docs/management/persistence.md:
```diff
@@ -259,26 +259,35 @@ and starts appending new data into the new file.
-The second CONFIG command is used to turn off snapshotting persistence. This is optional, if you wish you can take both the persistence methods enabled.
+* Enable AOF: `redis-cli config set appendonly yes`
+* Optionally disable RDB: `redis-cli config set save ""`
+* Wait for AOF rewrite to finish persisting the data. You can do that by
```
This is important to check only if you want to restart the server. It is not necessary to restart the server. Maybe move this to "Before restarting the server" section, or at least mention that it's related to restarting?

threema-danilo replied:
Good point, updated in 2f202c889f73918a2f377737b015d287ddf3debd

zuiderkwast commented on docs/management/persistence.md:
```diff
@@ -259,26 +259,35 @@ and starts appending new data into the new file.
+Next time you restart the server:
+
+* Make sure your database contains the same number of keys it contained previously.
+* Make sure writes are appended to the append only file correctly.
```
Writes are appended to AOF file as soon as AOF rewrite is finished. Restart is not necessary. It is more related to the bullet "Wait for AOF rewrite to finish ... INFO persistence ..." above, so maybe add this in the end of that bullet?

zuiderkwast reviewed (commented):
Better. Thanks.

* Please don't rebase and force-push. Use merge and regular push instead. It makes it easier to see what's changed since last review. We use squash-merge anyway in the end.
* Have a look at https://github.com/redis/redis-doc#styling-guidelines, especially "Start every sentence on a new line."

dwdougherty commented on docs/management/persistence.md:
```diff
@@ -274,20 +274,15 @@ Switch to AOF on live database:
-  the server will start again with the old configuration, resulting in a loss
-  of your data.)
+* Make sure writes are appended to the append only file correctly.
+* **IMPORTANT:** Update your `redis.conf` (potentially through `CONFIG REWRITE`) and ensure that it matches the config above.
```
```suggestion
* **IMPORTANT:** Update your `redis.conf` (potentially through `CONFIG REWRITE`) and ensure that it matches the configuration above.
```

dwdougherty commented on docs/management/persistence.md:
```diff
@@ -274,20 +274,15 @@ Switch to AOF on live database:
-  of your data.)
+* Make sure writes are appended to the append only file correctly.
+* **IMPORTANT:** Update your `redis.conf` (potentially through `CONFIG REWRITE`) and ensure that it matches the config above.
+  (If you forget this step, when you restart the server, the configuration changes will be lost and the server will start again with the old configuration, resulting in a loss of your data.)
```
```suggestion
  If you forget this step, when you restart the server, the configuration changes will be lost and the server will start again with the old configuration, resulting in a loss of your data.
```

dwdougherty commented on docs/management/persistence.md:
```diff
@@ -274,20 +274,15 @@ Switch to AOF on live database:
 Next time you restart the server:
 
+* Wait for AOF rewrite to finish persisting the data.
+  You can do that by watching `INFO persistence`, waiting for `aof_rewrite_in_progress` and `aof_rewrite_scheduled` to be `0`, and validate that `aof_last_bgrewrite_status` is `ok`.
```
```suggestion
  You can do that by watching `INFO persistence`, waiting for `aof_rewrite_in_progress` and `aof_rewrite_scheduled` to be `0`, and validating that `aof_last_bgrewrite_status` is `ok`.
```

dwdougherty reviewed (changes requested):
Hopefully you don't a tiny bit of copy-editing.

zuiderkwast commented on docs/management/persistence.md:
```diff
@@ -259,26 +259,30 @@ and starts appending new data into the new file.
-**IMPORTANT:** remember to edit your redis.conf to turn on the AOF, otherwise
-when you restart the server the configuration changes will be lost and the
-server will start again with the old configuration.
+* Wait for AOF rewrite to finish persisting the data.
```
```suggestion
* Before restarting the server, wait for AOF rewrite to finish persisting the data.
```

zuiderkwast commented on docs/management/persistence.md:
```diff
@@ -259,26 +259,30 @@ and starts appending new data into the new file.
-server will start again with the old configuration.
+* Wait for AOF rewrite to finish persisting the data.
+  You can do that by watching `INFO persistence`, waiting for `aof_rewrite_in_progress` and `aof_rewrite_scheduled` to be `0`, and validate that `aof_last_bgrewrite_status` is `ok`.
+* Make sure your database contains the same number of keys it contained previously.
```
```suggestion
* After restarting the server, check that your database contains the same number of keys it contained previously.
```

zuiderkwast reviewed (commented):
Some things to are important to check before restarting and something else to check after restarting. I added suggestions for those.

Don't worry about previous force-pushes. It's just good to know for future PRs, also in redis main repo.

threema-danilo commented:
Review feedback addressed! Let's hope that after 7 rounds of reviews by 5 people, the data loss prevention warning is now ready to merge 😄
